### PR TITLE
Adding the SecureBoot in the 'Before We Begin'

### PR DESCRIPTION
**Pull request by bbyykk**

### CODE REVIEW

File: lkmpg.tex [hunk 1]
```diff
@@ -174,6 +174,21 @@ \subsection{Before We Begin}
         You will not see it unless you look through your \sh|journalctl| .
         See \ref{sec:helloworld} for details.
         To have immediate access to this information, do all your work from the console.
+  \label{sec:using_x}
+  \item SecureBoot.
+        Many modern PC or laptop are shipped with UEFI SecureBoot enabled by default. It is a security standard that can make sure the device boots using only software that is trusted by original equipment manufacturer.
```
Comment (jserv) [inline lkmpg.tex:179]: Rephrase as following:
> Many contemporary computers are pre-configured with UEFI SecureBoot enabled.

File: lkmpg.tex [hunk 2]
```diff
@@ -174,6 +174,21 @@ \subsection{Before We Begin}
...
+		The default Linux kernel from some distributions have also enabled the SecureBoot. For such distribution, the kernel module has to be signed with the security key or you would get the "\emph{ERROR: could not insert module}" when you insert your first hello world module:
+\begin{codebash}
+insmod ./hello-1.ko
+\end{codebash}
+        And then you can check further with "dmesg"and see the following text:
```
Comment (linD026) [inline lkmpg.tex:184]: Add the space between "dmesg" and "and".

File: lkmpg.tex [hunk 3]
```diff
@@ -174,6 +174,21 @@ \subsection{Before We Begin}
...
+
+           \emph{Lockdown: insmod: unsigned module loading is restricted;
+		see man kernel lockdown.7}
+
+	If you got this message, the simplest way is to disable the UEFI SecureBoot from the PC/laptop boot menu to have your "hello world" to be inserted. Of course you can go through many complicated steps to generate keys, install keys to your system, and finally sign your module to make it work. However, this needs another chapter for the advanced users but not suitable for beginners.
```
Comment (linD026) [inline lkmpg.tex:189]: Split the sentences like this:
```
If you ... .
Of course, ... .
However, ... .
```

Comment (linD026) [inline lkmpg.tex:189]: Also, Indent properly.
I think you miss 4 spaces here.

File: lkmpg.tex [hunk 4]
```diff
@@ -174,6 +174,21 @@ \subsection{Before We Begin}
         You will not see it unless you look through your \sh|journalctl| .
         See \ref{sec:helloworld} for details.
         To have immediate access to this information, do all your work from the console.
+  \label{sec:using_x}
```
Comment (linD026) [inline lkmpg.tex:177]: I got the error.
```
Latexmk: List of undefined refs and citations:
  Label `sec:using_x' multiply defined
```
Please use another label or remove it.

File: lkmpg.tex (hunk 2 above)
Comment (linD026) [inline lkmpg.tex:184]: Use `\sh|dmesg|` instead of `"dmesg"` for consisting with other use cases.

File: lkmpg.tex [hunk 5]
```diff
@@ -174,6 +174,21 @@ \subsection{Before We Begin}
...
+  \item SecureBoot.
+        Many modern PC or laptop are shipped with UEFI SecureBoot enabled by default. It is a security standard that can make sure the device boots using only software that is trusted by original equipment manufacturer.
+		The default Linux kernel from some distributions have also enabled the SecureBoot. For such distribution, the kernel module has to be signed with the security key or you would get the "\emph{ERROR: could not insert module}" when you insert your first hello world module:
```
Comment (linD026) [inline lkmpg.tex:180]: One sentence one line, please.

**Review by linD026 — CHANGES_REQUESTED** [review]

**Review by linD026 — CHANGES_REQUESTED** [review]
Please use 4 spaces instead of the tab.

File: lkmpg.tex [hunk 6]
```diff
@@ -174,6 +174,24 @@ \subsection{Before We Begin}
...
+        For such distributions, the kernel module has to be signed with the security key or you would get the "\emph{ERROR: could not insert module}" when you insert your first hello world module:
+\begin{codebash}
+insmod ./hello-1.ko
+\end{codebash}
+        And then you can check further with \sh|dmesg| and see the following text:
+
+           \emph{Lockdown: insmod: unsigned module loading is restricted;
+                 see man kernel lockdown.7}
+
+        If you got this message, the simplest way is to disable the UEFI SecureBoot from the PC/laptop boot menu to have your "hello world" to be inserted.
```
Comment (jserv) [inline lkmpg.tex:190]: Replace "hello world" with `hello-1` module.

**Comment by linD026** [comment]
It looks good to me.
But I just wonder that if we write the enumerate part, does the codebash part still need to be indented?
like, here is the pull request used:
```tex
        For such distributions, the kernel module has to be signed with the security key or you would get the "\emph{ERROR: could not insert module}" when you insert your first hello world module:
\begin{codebash}
insmod ./hello-1.ko
\end{codebash}
        And then you can check further with \sh|dmesg| and see the following text:
```
image

For the others:
```tex
  \item {
  You can use the \verb|-E| flag to temporarily preserve them.
  
  \begin{codebash}
    $ sudo -E make -p | grep PWD
    PWD = /home/ubuntu/temp
    OLDPWD = /home/ubuntu
    echo $(PWD)
  \end{codebash}
  }
```
image

The difference is for the latter case the code will be indented in the pdf.
However, the former case looks better.

File: lkmpg.tex (hunk 6 above)
Comment (bbyykk) [reply lkmpg.tex:190]: > Replace "hello world" with `hello-1` module.

will do

File: lkmpg.tex [hunk 7]
```diff
@@ -174,6 +174,25 @@ \subsection{Before We Begin}
...
+insmod ./hello-1.ko
+\end{codebash}
+        And then you can check further with \sh|dmesg| and see the following text:
+
+           \emph{Lockdown: insmod: unsigned module loading is restricted;
+                 see man kernel lockdown.7}
+
+        If you got this message, the simplest way is to disable the UEFI SecureBoot from the PC/laptop boot menu to have your "hello-1" to be inserted.
+        Of course you can go through complicated steps to generate keys, install keys to your system, and finally sign your module to make it work.
+        However, this needs another chapter for the advanced users but not suitable for beginners.
```
Comment (jserv) [inline lkmpg.tex:193]: Add some advanced materials as a reference.

File: lkmpg.tex [hunk 8]
```diff
@@ -190,7 +190,9 @@ \subsection{Before We Begin}
 
         If you got this message, the simplest way is to disable the UEFI SecureBoot from the PC/laptop boot menu to have your "hello-1" to be inserted.
         Of course you can go through complicated steps to generate keys, install keys to your system, and finally sign your module to make it work.
-        However, this needs another chapter for the advanced users but not suitable for beginners.
+        However, this is not suitable for beginners.
+        Please follow the steps in \href{https://ubuntu.com/blog/how-to-sign-things-for-secure-boot}{how-to-sign-things-for-secure-boot} if you are interested.
```
Comment (jserv) [inline lkmpg.tex:194]: Don't change the title of a specified source. That is, you shall remain "How to sign things for Secure Boot" for `\href`.

File: lkmpg.tex [hunk 9]
```diff
@@ -190,7 +190,9 @@ \subsection{Before We Begin}
...
+        And you can learn more with \href{https://docs.oracle.com/en/operating-systems/oracle-linux/secure-boot/}{Oracle® Linux Working With UEFI Secure Boot}
```
Comment (jserv) [inline lkmpg.tex:195]: Regarding the content, Debian Wiki might be a better alternative. See https://wiki.debian.org/SecureBoot

Comment (bbyykk) [reply lkmpg.tex:195]: Agreed. I went through it and it is easier to read and follow the instructions

**Review by jserv — CHANGES_REQUESTED** [review]
Rebase the latest `master` branch.

**Comment by jserv** [comment]
GitHub Action [#177](https://github.com/sysprog21/lkmpg/actions/runs/3367433288/jobs/5628664663) reported as following:
> Smatch failed: 1 warning(s), 0 error(s)
> /home/runner/work/lkmpg/lkmpg/examples/vinput.c:372 vinput_init() warn: missing error code 'err'

It seems to be false alert, and we shall avoid. @linD026, do you have any idea to make progress with static analysis?

**Comment by linD026** [comment]
> GitHub Action [#177](https://github.com/sysprog21/lkmpg/actions/runs/3367433288/jobs/5628664663) reported as following:
> 
> > Smatch failed: 1 warning(s), 0 error(s)
> > /home/runner/work/lkmpg/lkmpg/examples/vinput.c:372 vinput_init() warn: missing error code 'err'
> 
> It seems to be false alert, and we shall avoid. @linD026, do you have any idea to make progress with static analysis?

I open the PR (https://github.com/sysprog21/lkmpg/pull/176).
I think this would fix the report.

**Comment by jserv** [comment]
> > GitHub Action [#177](https://github.com/sysprog21/lkmpg/actions/runs/3367433288/jobs/5628664663) reported as following:
> > It seems to be false alert, and we shall avoid. @linD026, do you have any idea to make progress with static analysis?
> 
> I open the PR (#176). I think this would fix the report.

Thank @linD026 for quick-fix. @bbyykk, please rebase again to pass CI pipeline.

**Pull request closed without merging**